From 6232970736d9d23a2c9534cd7217c8554be67745 Mon Sep 17 00:00:00 2001
From: PulkoMandy <pulkomandy@pulkomandy.tk>
Date: Sat, 21 Jun 2025 20:21:15 +0200
Subject: Get it to build on Haiku

- Disable search for ucontext.h on Haiku (Haiku defines ucontext struct in signal.h)
- Search for network functions in libnetwork
- Change libbsd include paths: Haiku's libbsd needs the normal <string.h> include
  instead of <bsd/string.h> (otherwise there are problems with the use
  of #include_next in Haiku headers)
- Disable code for TUN/TAP handling (Haiku specific code needs to be
  written for this): _IOW and _IOR don't exist on Haiku, and the TUN configration uses a different
  interface. TODO use configure script to detect the needed flags
  instead of adding Haiku special cases in the sourcecode.
- Use elf_i386_haiku instead of elf_i386 (not sure why Haiku's binutils don't know about elf_i386)
- Implement CPU features detection (there is no /proc/cpuinfo in Haiku; use cpuid and
  get_cpu_topology_info instead)
- Enable definition of getmcontext and setmcontext for Haiku
- In Haiku, the libexec directory is /system/lib and the lib directory is /system/develop/lib. This
  should be handled using LIBDIR and LIBEXECDIR ideally (just like DATADIR and other directories
  handled by autoconf)
- Set MAP_NORESERVE and B_OVERCOMMITING_AREA where needed (in Linux this is implied by
  PROT_NONE). Still not working and causing a KDL, see
  http://dev.haiku-os.org/ticket/19264

diff --git a/configure.ac b/configure.ac
index 6b51284..511a7df 100644
--- a/configure.ac
+++ b/configure.ac
@@ -505,6 +505,12 @@ AC_CHECK_HEADERS([asm/ucontext.h],,, [
   #include <ucontext.h>
   #include <signal.h>
 ])
+
+AC_SEARCH_LIBS([accept], [network],[],[
+  AC_MSG_ERROR([Unable t find BSD sockets support])
+])
+
+if test "$CONFIG_OS" != "Haiku"; then
 if test "$machine" != "wasm32"; then
   AC_CHECK_HEADERS([ucontext.h],, [
     AS_UNSET([ac_cv_header_ucontext_h])
@@ -514,6 +520,7 @@ if test "$machine" != "wasm32"; then
     #define _XOPEN_SOURCE 500
   ])])
 fi
+fi
 AC_CHECK_HEADERS([linux/signal.h],,, [#include <signal.h>])
 AC_CHECK_HEADERS([sys/soundcard.h], [
   USE_OSS=1
@@ -844,7 +851,7 @@ fi
 DOSEMU_CFLAGS="${DOSEMU_CFLAGS} ${OPT} ${PIPE}"
 DOSEMU_CPPFLAGS="${DOSEMU_CPPFLAGS} -MD -DCFLAGS_STR=\"$DOSEMU_CFLAGS $CFLAGS\""
 DOSEMU_VERSION=`cd $srcdir && ./getversion -b`
-AS_LDFLAGS="-melf_i386"
+AS_LDFLAGS="-melf_i386_haiku"
 
 AC_SUBST(XASFLAGS)
 AC_SUBST(ASFLAGS)
diff --git a/src/base/init/config.c b/src/base/init/config.c
index 0d5f8dd..7b21340 100644
--- a/src/base/init/config.c
+++ b/src/base/init/config.c
@@ -43,6 +43,14 @@
 #include "mhpdbg.h"
 #include "mapping/mapping.h"
 
+#ifdef __HAIKU__
+/* Rename some conflicting functions that we don't need */
+#define read_port haiku_read_port
+#define write_port haiku_write_port
+#include <kernel/OS.h>
+#undef read_port
+#undef write_port
+#endif
 /*
  * Options used in config_init().
  *
@@ -812,10 +820,84 @@ void secure_option_preparse(int *argc, char **argv)
 
 static void read_cpu_info(void)
 {
-    char *cpuflags, *cpu;
     int k = 3;
+    int cores = 0;
     int err;
 
+#ifdef __HAIKU__
+    cpuid_info info;
+
+    err = get_cpuid(&info, 1, 0);
+    if (err != B_OK)
+        return;
+
+    k = info.eax_1.family;
+    bool fpu = info.eax_1.extended_features & (1 << 0);
+    bool tsc = info.eax_1.extended_features & (1 << 4);
+    bool mmx = info.eax_1.extended_features & (1 << 23);
+    bool fxsr = info.eax_1.extended_features & (1 << 24);
+    bool sse = info.eax_1.extended_features & (1 << 25);
+
+    cpu_topology_node_info topology[128];
+    int count = 128;
+    get_cpu_topology_info(topology, &count);
+
+    if (k >= 5) {
+        config.realcpu = CPU_586;
+
+#ifdef X86_EMULATOR
+        if (mmx || sse) {
+	  config.cpuprefetcht0 = 1;
+        }
+#endif
+#ifdef __i386__
+        if (fxsr) {
+          config.cpufxsr = 1;
+	  if (sse)
+	    config.cpusse = 1;
+	}
+#endif
+        if (tsc) {
+            long long chz = 0;
+            for (int i = 0; i < count; i++) {
+                if (topology[i].type == B_TOPOLOGY_CORE) {
+                    chz = topology[i].data.core.default_frequency;
+                    cores ++;
+                }
+            }
+		/* speed division factor to get 1us from CPU clock */
+		config.cpu_spd = (LLF_US*1000000)/chz;
+
+		/* speed division factor to get 838ns from CPU clock */
+		config.cpu_tick_spd = (LLF_TICKS*1000000)/chz;
+
+		config.CPUSpeedInMhz = chz / 1000000;
+
+        }
+    } else if (k == 4) {
+        config.realcpu = CPU_486;
+    } else if (k != 3) {
+        error("Unknown CPU type!\n");
+    }
+
+    if (config.mathco) {
+        config.mathco = fpu;
+    }
+
+    err = get_cpuid(&info, 9, 0);
+    if (err != B_OK)
+      return;
+
+    if (info.regs.ecx & (1 << 2))
+      config.umip = 1;
+    else
+      warn("Your CPU doesn't support UMIP\n");
+
+    if (cores > 0)
+        config.smp = 1;
+#else
+    char *cpuflags, *cpu;
+
     err = open_proc_scan("/proc/cpuinfo");
     if (err)
       return;
@@ -853,6 +935,7 @@ static void read_cpu_info(void)
 	    config.cpusse = 1;
 	}
 #endif
+
         if (cpuflags && strstr(cpuflags, "tsc")) {
           /* bogospeed currently returns 0; should it deny
            * pentium features, fall back into 486 case */
@@ -921,6 +1004,7 @@ static void read_cpu_info(void)
       config.smp = 1;		/* for checking overrides, later */
     }
     close_proc_scan();
+#endif
 }
 
 static void config_post_process(void)
diff --git a/src/base/lib/mapping/mapping.c b/src/base/lib/mapping/mapping.c
index 21bbe0a..e21b1ed 100644
--- a/src/base/lib/mapping/mapping.c
+++ b/src/base/lib/mapping/mapping.c
@@ -42,6 +42,14 @@
 #include <linux/version.h>
 #endif
 
+#ifdef __HAIKU__
+#define read_port haiku_read_port
+#define write_port write_read_port
+#include <private/system/vm_defs.h>
+#undef read_port
+#undef write_port
+#endif
+
 enum { MEM_BASE, KVM_BASE, JIT_BASE,
 #ifdef __i386__
        VM86_BASE,
@@ -413,7 +421,7 @@ static void *do_huge_page(int flags, size_t mapsize, int protect)
 
 void *mmap_mapping_huge_page_aligned(int cap, size_t mapsize, int protect)
 {
-  int flags = (cap & MAPPING_INIT_LOWRAM) ? _MAP_32BIT : 0;
+  int flags = (cap & MAPPING_INIT_LOWRAM) ? _MAP_32BIT | MAP_NORESERVE : 0;
   void *addr = do_huge_page(flags, mapsize, protect);
   if (addr == MAP_FAILED)
     return addr;
@@ -784,7 +792,7 @@ void *alloc_mapping_huge_page_aligned(int cap, size_t mapsize)
 {
   void *addr;
   Q__printf("MAPPING: alloc_huge_page_aligned, cap=%s size=%#zx\n", cap, mapsize);
-  addr = do_huge_page(0, mapsize, PROT_NONE);
+  addr = do_huge_page(0, mapsize, PROT_NONE | B_OVERCOMMITTING_AREA);
   return addr == MAP_FAILED ? MAP_FAILED : do_alloc_mapping(cap, mapsize, addr);
 }
 
diff --git a/src/base/lib/mcontext/asm.SS b/src/base/lib/mcontext/asm.SS
index fefe877..bf1dba5 100644
--- a/src/base/lib/mcontext/asm.SS
+++ b/src/base/lib/mcontext/asm.SS
@@ -12,6 +12,18 @@
 #endif
 #endif
 
+#if defined(__HAIKU__) && defined(__i386__)
+#define NEEDX86CONTEXT 1
+#define SET setmcontext
+#define GET getmcontext
+#else
+#if defined(__HAIKU__) && defined(__x86_64__)
+#define NEEDAMD64CONTEXT 1
+#define SET _setmcontext
+#define GET _getmcontext
+#endif
+#endif
+
 #if defined(__OpenBSD__) && defined(__i386__)
 #define NEEDX86CONTEXT 1
 #define SET setmcontext
diff --git a/src/base/misc/utilities.c b/src/base/misc/utilities.c
index 087561a..b2f4774 100644
--- a/src/base/misc/utilities.c
+++ b/src/base/misc/utilities.c
@@ -598,7 +598,7 @@ char *path_prefix(const char *suffix)
         ret = assemble_path(p, suffix);
       else if (strcmp(p1 + 1, "dosemu2") == 0) {
         p1 = strrchr(p, '/');
-        if (p1 && strcmp(p1 + 1, "libexec") == 0) {
+        if (p1 && strcmp(p1 + 1, "lib") == 0) {
           *p1 = '\0';
           ret = assemble_path(p, suffix);
         }
diff --git a/src/dosemu b/src/dosemu
index d423af5..176f715 100644
--- a/src/dosemu
+++ b/src/dosemu
@@ -23,7 +23,7 @@ get_binary() {
     fi
 
   else
-    BINARY="$prefix"/libexec/dosemu2/dosemu2.bin
+    BINARY="$prefix"/lib/dosemu2/dosemu2.bin
     if [ ! -f "$BINARY" ]; then
       echo "$BINARY does not exist"
       exit 1
diff --git a/src/dosext/net/libpacket.c b/src/dosext/net/libpacket.c
index a2a359f..66130f9 100644
--- a/src/dosext/net/libpacket.c
+++ b/src/dosext/net/libpacket.c
@@ -423,6 +423,9 @@ int GetDeviceMTU(void)
 
 int tun_alloc(char *dev)
 {
+#ifdef __HAIKU__
+      return -1;
+#else
       PRIV_SAVE_AREA
       struct ifreq ifr;
       int fd, err;
@@ -459,6 +462,7 @@ int tun_alloc(char *dev)
       strcpy(dev, ifr.ifr_name);
 
       return fd;
+#endif
 }
 
 static ssize_t pkt_read_eth(int pkt_fd, void *buf, size_t count)
diff --git a/src/dosext/net/tcp.c b/src/dosext/net/tcp.c
index 8742deb..69e1736 100644
--- a/src/dosext/net/tcp.c
+++ b/src/dosext/net/tcp.c
@@ -27,8 +27,13 @@
 #include "Netinet/tcp.h"
 #include <limits.h>
 #ifdef HAVE_LIBBSD
+#ifdef __HAIKU__
+/* Let Haiku headers do their own thing with #include_next... */
+#include <string.h>
+#else
 #include <bsd/string.h>
 #endif
+#endif
 #include "emu.h"
 #include "hlt.h"
 #include "int.h"
diff --git a/src/include/Linux/if_tun.h b/src/include/Linux/if_tun.h
index 5feacf5..16f9184 100644
--- a/src/include/Linux/if_tun.h
+++ b/src/include/Linux/if_tun.h
@@ -26,6 +26,7 @@
 #define TUN_TYPE_MASK   0x000f
 
 /* Ioctl defines */
+#ifndef __HAIKU__
 #define TUNSETNOCSUM  _IOW('T', 200, int) 
 #define TUNSETDEBUG   _IOW('T', 201, int) 
 #define TUNSETIFF     _IOW('T', 202, int) 
@@ -54,6 +55,7 @@
  */
 #define TUNSETVNETBE _IOW('T', 222, int)
 #define TUNGETVNETBE _IOR('T', 223, int)
+#endif
 
 /* TUNSETIFF ifr flags */
 #define IFF_TUN		0x0001
diff --git a/src/include/dosemu_config.h b/src/include/dosemu_config.h
index 0b18a0d..bbd7a8c 100644
--- a/src/include/dosemu_config.h
+++ b/src/include/dosemu_config.h
@@ -53,7 +53,11 @@ extern int config_check_only;
 #define  FREEDOS_DIR        "freedos"         /* freedos dir name */
 #define  FDBOOT_DIR         "fdboot"          /* freedos boot dir name */
 #define  XBAT_DIR           "bat"             /* extras */
+#ifdef __HAIKU__
+#define  DOSEMULIB_DEFAULT  "data/dosemu"
+#else
 #define  DOSEMULIB_DEFAULT  "share/dosemu"
+#endif
 #define  DOSEMUCMDS_DEFAULT DOSEMULIB_DEFAULT "/" CMDS_SUFF
 #define  DOSEMUEXEC_DEFAULT  LIBEXECDIR "/dosemu"
 #define  DOSEMUIMAGE_DEFAULT "~/" LOCALDIR_BASE_NAME
diff --git a/src/plugin/fdpp/hooks.c b/src/plugin/fdpp/hooks.c
index d2f0e9f..1a420ea 100644
--- a/src/plugin/fdpp/hooks.c
+++ b/src/plugin/fdpp/hooks.c
@@ -20,8 +20,12 @@
  * Author: Stas Sergeev
  */
 #ifdef HAVE_LIBBSD
+#ifdef __HAIKU__
+#include <string.h>
+#else
 #include <bsd/string.h>
 #endif
+#endif
 #include <stddef.h>
 #include <fdpp/loader.h>
 #include <fdpp/thunks.h>
-- 
2.48.1

